src/main/java/org/broadinstitute/hellbender/utils/spark/SparkUtils.java

```diff
@@ -103,4 +123,358 @@ public static boolean pathExists(final JavaSparkContext ctx, final Path targetPa
+            if (partitionIndexes.length == 0) {
+                // interval does not overlap any partition - skip it
+                continue;
+            }
+            Arrays.sort(partitionIndexes);
+            int startIndex = partitionIndexes[0];
+            int endIndex = partitionIndexes[partitionIndexes.length - 1];
+            indexedIntervals.add(new PartitionLocatable<I>(startIndex, interval));
+            if (endIndex > maxEndPartitionIndexes.get(startIndex)) {
+                maxEndPartitionIndexes.set(startIndex, endIndex);
+            }
+        }
+
+        JavaRDD<R> coalescedRdd = coalesce(reads, readClass, new RangePartitionCoalescer(maxEndPartitionIndexes));
+
+        // Create an RDD of intervals with the same number of partitions as the reads, and where each interval
+        // is in its start partition.
+        JavaRDD<I> intervalsRdd = ctx.parallelize(indexedIntervals)
+                .mapToPair(interval ->
+                        new Tuple2<>(interval.getPartitionIndex(), interval.getLocatable()))
+                .partitionBy(new KeyPartitioner(reads.getNumPartitions())).values();
+
+        // zipPartitions on coalesced read partitions and intervals, and apply the function f
+        return coalescedRdd.zipPartitions(intervalsRdd, f);
+    }
+
+    /**
+     * Turn a pair of iterators over intervals and reads, into a single iterator over pairs made up of an interval and
+     * the reads that overlap it. Intervals with no overlapping reads are dropped.
+     */
+    static <R extends Locatable, I extends Locatable> Iterator<Tuple2<I, Iterable<R>>> readsPerShard(Iterator<R> reads, Iterator<I> shards, SAMSequenceDictionary sequenceDictionary) {
+        if (!shards.hasNext()) {
+            return Collections.emptyIterator();
+        }
+        PeekingIterator<R> peekingReads = Iterators.peekingIterator(reads);
+        PeekingIterator<I> peekingShards = Iterators.peekingIterator(shards);
+        Iterator<Tuple2<I, Iterable<R>>> iterator = new AbstractIterator<Tuple2<I, Iterable<R>>>() {
+            // keep track of current and next, since reads can overlap two shards
+            I currentShard = peekingShards.next();
+            I nextShard = peekingShards.hasNext() ? peekingShards.next() : null;
+            List<R> currentReads = Lists.newArrayList();
+            List<R> nextReads = Lists.newArrayList();
+
+            @Override
+            protected Tuple2<I, Iterable<R>> computeNext() {
+                if (currentShard == null) {
+                    return endOfData();
+                }
+                while (peekingReads.hasNext()) {
+                    if (toRightOf(currentShard, peekingReads.peek(), sequenceDictionary)) {
+                        break;
+                    }
+                    R read = peekingReads.next();
+                    if (overlaps(currentShard, read)) {
+                        currentReads.add(read);
```

Comment:
Yes, it might be. There should be a way to downsample - even skip over partitions that have the same reads in them - but this could be tackled as a follow-on piece of work.